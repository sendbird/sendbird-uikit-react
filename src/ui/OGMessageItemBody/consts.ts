export const OG_MESSAGE_BODY_CLASSNAME = 'sendbird-og-message-item-body__text-bubble';
