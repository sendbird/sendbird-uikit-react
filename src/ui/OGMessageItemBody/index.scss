@import '../../styles/variables';

.sendbird-og-message-item-body {
  position: relative;
  display: inline-block;
  // TODO: Change fixing width and height lengths
  width: 100%;
  min-width: 320px;
  max-width: 400px;
  @include mobile() {
    width: 66vw;
    min-width: 160px;
  }

  .sendbird-og-message-item-body__text-bubble {
    position: relative;
    padding: 8px 12px;
    box-sizing: border-box;
    border-radius: 16px 16px 0px 0px;
    word-break: break-word;
    white-space: pre-line;

    .sendbird-og-message-item-body__text-bubble__message {
      display: inline;
      margin: 0px 2px;
    }
  }

  .sendbird-og-message-item-body__og-thumbnail {
    position: relative;
    overflow: hidden;
    .sendbird-og-message-item-body__og-thumbnail__image {
      width: 100%;
      position: relative;
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }

  &.reactions .sendbird-og-message-item-body__description {
    border-radius: 0px;
  }
  .sendbird-og-message-item-body__description {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    padding: 8px 12px;
    display: flex;
    flex-direction: column;
    border-radius: 0px 0px 16px 16px;
    @include themed() {
      background-color: t(bg-1);
    }
    .sendbird-og-message-item-body__description__title,
    .sendbird-og-message-item-body__description__description {
      margin-bottom: 4px;
      width: 100%;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      display: inline-block;
    }
    .sendbird-og-message-item-body__description__url {
      width: 100%;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      display: inline-block;
      margin: 4px 0px;
    }
  }

  .sendbird-og-message-item-body__og-thumbnail:hover,
  .sendbird-og-message-item-body__description:hover {
    cursor: pointer;
  }

  &.outgoing {
    .sendbird-og-message-item-body__text-bubble {
      @include themed() {
        background-color: t(primary-3);
        color: t(on-content-1);
      }
    }
  }
  &.incoming {
    .sendbird-og-message-item-body__text-bubble {
      @include themed() {
        background-color: t(bg-1);
        color: t(on-bg-1);
      }
    }
  }
  &.mouse-hover, &:hover {
    .sendbird-og-message-item-body__description {
      @include themed() { background-color: t(bg-2) }
    }
    &.outgoing {
      .sendbird-og-message-item-body__text-bubble {
        @include themed() { background-color: t(primary-4) }
      }
    }
    &.incoming {
      .sendbird-og-message-item-body__text-bubble {
        @include themed() { background-color: t(bg-2) }
      }
    }
  }

  .sendbird-og-message-item-body__cover {
    display: none;
    position: absolute;
    top: 0%;
    left: 0;
    width: 100%;
    height: 100%;
    border-radius: 16px 16px 0px 0px;
  }
}

div.sendbird-og-message-item-body__text-bubble {
  margin: 0px;
}
