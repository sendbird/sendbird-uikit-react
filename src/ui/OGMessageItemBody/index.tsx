import './index.scss';
import React, { ReactElement, useContext, useMemo } from 'react';
import type { UserMessage } from '@sendbird/chat/message';

import Word from '../Word';
import ImageRenderer from '../ImageRenderer';
import Icon, { IconTypes } from '../Icon';
import Label, { LabelTypography, LabelColors } from '../Label';
import uuidv4 from '../../utils/uuid';
import { getClassName, isEditedMessage } from '../../utils';
import { LocalizationContext } from '../../lib/LocalizationContext';

interface Props {
  className?: string | Array<string>;
  message: UserMessage;
  isByMe?: boolean;
  mouseHover?: boolean;
  isMentionEnabled?: boolean;
  isReactionEnabled?: boolean;
}

export default function OGMessageItemBody({
  className,
  message,
  isByMe = false,
  mouseHover = false,
  isMentionEnabled = false,
  isReactionEnabled = false,
}: Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const openOGUrl = (): void => {
    if (message?.ogMetaData?.url) window.open(message?.ogMetaData?.url);
  };
  const isMessageMentioned = isMentionEnabled && message?.mentionedMessageTemplate?.length > 0 && message?.mentionedUsers?.length > 0;
  const sentences: Array<Array<string>> = useMemo(() => {
    if (isMessageMentioned) {
      return message?.mentionedMessageTemplate?.split(/\n/).map((sentence) => sentence.split(/\s/));
    }
    return message?.message?.split(/\n/).map((sentence) => sentence.split(/\s/));
  }, [message?.mentionedMessageTemplate]);
  return (
    <div className={getClassName([
      className,
      'sendbird-og-message-item-body',
      isByMe ? 'outgoing' : 'incoming',
      mouseHover ? 'mouse-hover' : '',
      (isReactionEnabled && message?.reactions?.length > 0) ? 'reactions' : '',
    ])}>
      <Label
        key={uuidv4()}
        type={LabelTypography.BODY_1}
        color={isByMe ? LabelColors.ONCONTENT_1 : LabelColors.ONBACKGROUND_1}
      >
        <div className="sendbird-og-message-item-body__text-bubble">
          {
            sentences?.map((sentence, index) => {
              return [
                sentence?.map((word) => {
                  return (
                    <Word
                      key={uuidv4()}
                      word={word}
                      message={message}
                      isByMe={isByMe}
                    />
                  );
                }),
                sentences?.[index + 1] ? <br /> : null,
              ]
            })
          }
          {
            isEditedMessage(message) && (
              <Label
                className="sendbird-og-message-item-body__text-bubble__message"
                type={LabelTypography.BODY_1}
                color={isByMe ? LabelColors.ONCONTENT_2 : LabelColors.ONBACKGROUND_2}
              >
                {` ${stringSet.MESSAGE_EDITED} `}
              </Label>
            )
          }
        </div>
      </Label>
      <div
        className="sendbird-og-message-item-body__og-thumbnail"
        onClick={openOGUrl}
      >
        <ImageRenderer
          className="sendbird-og-message-item-body__og-thumbnail__image"
          url={message?.ogMetaData?.defaultImage?.url || ''}
          alt={message?.ogMetaData?.defaultImage?.alt}
          // TODO: Change fixing width and height lengths
          defaultComponent={(
            <div className="sendbird-og-message-item-body__og-thumbnail__place-holder">
              <Icon
                className="sendbird-og-message-item-body__og-thumbnail__place-holder__icon"
                type={IconTypes.THUMBNAIL_NONE}
                width="56px"
                height="56px"
              />
            </div>
          )}
        />
      </div>
      <div
        className="sendbird-og-message-item-body__description"
        onClick={openOGUrl}
      >
        {message?.ogMetaData?.title && (
          <Label
            className="sendbird-og-message-item-body__description__title"
            type={LabelTypography.SUBTITLE_2}
            color={LabelColors.ONBACKGROUND_1}
          >
            {message.ogMetaData.title}
          </Label>
        )}
        {message?.ogMetaData?.description && (
          <Label
            className="sendbird-og-message-item-body__description__description"
            type={LabelTypography.BODY_2}
            color={LabelColors.ONBACKGROUND_1}
          >
            {message.ogMetaData.description}
          </Label>
        )}
        {message?.ogMetaData?.url && (
          <Label
            className="sendbird-og-message-item-body__description__url"
            type={LabelTypography.CAPTION_3}
            color={LabelColors.ONBACKGROUND_2}
          >
            {message.ogMetaData.url}
          </Label>
        )}
      </div>
      <div className="sendbird-og-message-item-body__cover" />
    </div>
  );
}
