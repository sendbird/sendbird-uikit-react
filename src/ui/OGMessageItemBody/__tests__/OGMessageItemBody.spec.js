import React from 'react';
import { render } from '@testing-library/react';

import OGMessageItemBody from "../index";

describe('ui/OGMessageItemBody', () => {
  it('should do a snapshot test of the OGMessageItemBody DOM', function() {
    const text = "example-text";
    const { asFragment } = render(<OGMessageItemBody />);
    expect(asFragment()).toMatchSnapshot();
  });
});
