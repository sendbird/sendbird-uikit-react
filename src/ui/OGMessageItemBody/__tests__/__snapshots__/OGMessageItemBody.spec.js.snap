// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OGMessageItemBody should do a snapshot test of the OGMessageItemBody DOM 1`] = `
<div
  className=" sendbird-og-message-item-body incoming  "
>
  <span
    className="sendbird-label sendbird-label--body-1 sendbird-label--color-onbackground-1"
  >
    <p
      className="sendbird-og-message-item-body__text-bubble"
    />
  </span>
  <div
    className="sendbird-og-message-item-body__og-thumbnail"
    onClick={[Function]}
  >
    <div
      className="sendbird-og-message-item-body__og-thumbnail__image sendbird-image-renderer"
      style={
        Object {
          "height": "180px",
          "maxWidth": "400px",
          "minWidth": "320px",
          "width": "100%",
        }
      }
    >
      <div
        className="sendbird-image-renderer__image"
        style={
          Object {
            "backgroundImage": "url()",
            "backgroundPosition": "center",
            "backgroundRepeat": "no-repeat",
            "backgroundSize": "cover",
            "borderRadius": null,
            "height": "180px",
            "maxWidth": "400px",
            "minWidth": "320px",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <img
        alt=""
        className="sendbird-image-renderer__hidden-image-loader"
        onError={[Function]}
        onLoad={[Function]}
        src=""
      />
    </div>
  </div>
  <div
    className="sendbird-og-message-item-body__description"
    onClick={[Function]}
  />
  <div
    className="sendbird-og-message-item-body__cover"
  />
</div>
`;
